src/fetch/rewards.py

```diff
@@ -0,0 +1,167 @@
+"""Functionality for computing solver rewards
+
+The only function used in other parts of the code is `compute_solver_rewards` for generating
+aggregate reward information per solver."""
+
+from pandas import DataFrame, Series
+from src.fetch.prices import TokenConversion
+
+EPSILON_UPPER = 12_000_000_000_000_000
+EPSILON_LOWER = 10_000_000_000_000_000
+PERIOD_BUDGET_COW = 250000 * 10**18
+CONSISTENCY_REWARD_CAP_ETH = 6 * 10**18
+QUOTE_REWARD_COW = 6 * 10**18
+QUOTE_REWARD_CAP_ETH = 6 * 10**14
+
+
+def compute_solver_rewards(
+    batch_data_df: DataFrame, trade_data_df: DataFrame, converter: TokenConversion
+) -> DataFrame:
+    """Compute solver rewards
+    The input dataframes `batch_data_df` and `trade_data_df` containd data on batches and
+    trades, respectively. It requires TokenConversion to convert betwenn ETH and COW.
+    The function creates a new dataframe with information on primary, secondary, and quote rewards.
+    """
+    # primary reward
+    primary_rewards_df = compute_primary_rewards(batch_data_df, converter)
+
+    # secondary rewards
+    secondary_rewards_df = compute_secondary_rewards_df(
+        batch_data_df, primary_rewards_df, converter
+    )
+
+    # quote rewards
+    quote_rewards_df = compute_quote_rewards(trade_data_df, converter)
+
+    # merge different rewards
+    solver_rewards_df = (
+        primary_rewards_df.merge(secondary_rewards_df, on="solver", how="outer")
+    ).merge(quote_rewards_df, on="solver", how="outer")
+
+    return solver_rewards_df
+
+
+def compute_primary_rewards(
+    batch_data_df: DataFrame, converter: TokenConversion
+) -> DataFrame:
+    """Compute primary rewards from batch data"""
+    primary_rewards_df = (
+        batch_data_df.apply(compute_primary_reward_datum, axis=1, result_type="expand")
+        .rename(columns={0: "solver", 1: "primary_reward_eth"})
+        .astype(object)
+        .groupby("solver")
+        .sum()
+    )
+    primary_rewards_df["primary_reward_cow"] = (
+        primary_rewards_df["primary_reward_eth"]
+        .apply(converter.eth_to_token)
+        .astype(object)
+    )
+    return primary_rewards_df
+
+
+def compute_primary_reward_datum(row: Series) -> tuple[str, int]:
+    """Compute the reward of an auction from the winning and references scores"""
+    solver = row["winning_solver"]
+    if row["tx_hash"] is None:
+        reward = -min(row["reference_score"], EPSILON_LOWER)
+    else:
+        reward = min(row["winning_score"] - row["reference_score"], EPSILON_UPPER)
+
+    return solver, reward
+
+
+def compute_secondary_rewards_df(
+    batch_data_df: DataFrame, primary_rewards_df: DataFrame, converter: TokenConversion
+) -> DataFrame:
+    """Compute secondary rewards from batch data and primary rewards"""
+    secondary_allocation = max(
```

Comment:
I guess i was thinking to have this outside of the function, and just pass a budget to this one, but your approach is also fine.

Reply:
I could make the computation of the budget its own function and just use the result here.
I do not want to move any computation (if not super essential) out of the function since the fact that there is a budget and how it depends on primary rewards is a fundamental part of how secondary rewards are computed.

Reply:
ok let's keep it as is